lib/services/local/headlessauthn_watcher_test.go

```diff
@@ -114,40 +114,39 @@ func TestHeadlessAuthenticationWatcher(t *testing.T) {
 		waitCtx, waitCancel := context.WithTimeout(ctx, time.Second*2)
 		defer waitCancel()
 
-		// Create two waiters - a blocked consumer and a free consumer.
-		blockWait := make(chan struct{})
-		_, errC := waitInGoroutine(waitCtx, t, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
-			<-blockWait
-			return false, nil
-		})
-
+		// Create a waiter that we can block/unblock.
 		notifyReceived := make(chan struct{}, 1)
-		waitInGoroutine(waitCtx, t, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
+		blockWaiter := make(chan struct{})
+		_, errC := waitInGoroutine(waitCtx, t, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
 			notifyReceived <- struct{}{}
+			<-blockWaiter
 			return false, nil
 		})
 
-		// Create stub and wait for it to be caught by the free waiter.
+		// Create stub and wait for it to be caught by the waiter.
 		stub, err := identity.CreateHeadlessAuthenticationStub(ctx, pubUUID)
 		require.NoError(t, err)
 		<-notifyReceived
 
+		// perform a put to mark the blocked waiter as stale and
 		replace := *stub
 		replace.PublicKey = []byte(sshPubKey)
 		replace.User = "user"
-
-		// perform a put to mark the blocked waiter as stale and
-		// wait for it to be caught by the free waiter.
 		_, err = identity.CompareAndSwapHeadlessAuthentication(ctx, stub, &replace)
 		require.NoError(t, err)
-		<-notifyReceived
 
-		// delete the headless authentication and unblock.
+		require.Eventually(t, func() bool {
+			ok, err := w.CheckWaiterStale(pubUUID)
+			require.NoError(t, err)
```

Comment:
I don't think it's safe to call `NoError` inside a `require.Eventually`.

Reply:
@zmb3 Why? `require.Eventually` runs on the same goroutine, so the error will be propagated. Another alternative is 
```
if err != nil {
  return false
}
```
but in this case, you're losing the error message and keep checking until the deadline.

Reply:
Does it run on the same goroutine?

Reply:
No - it doesn't: https://github.com/stretchr/testify/blob/f36bfe3c337aa95c86f04c721acdbafb5ffb1611/assert/assertions.go#L1737